src/app/campaign/report/campaign-report.component.ts

```diff
@@ -0,0 +1,78 @@
+import { Component, Input, ChangeDetectionStrategy } from '@angular/core';
+import { DomSanitizer, SafeUrl } from '@angular/platform-browser';
+import { FlightState, Flight } from '../store/models';
+import moment from 'moment';
+
+@Component({
+  selector: 'grove-campaign-report',
+  template: `
+    <a [href]="reportDataCsv" download="{{ reportFilename }}.csv" mat-menu-item>Download CSV</a>
+  `,
+  changeDetection: ChangeDetectionStrategy.OnPush
+})
+export class CampaignReportComponent {
+  @Input() campaignName: string;
+  @Input() flights: FlightState[];
+  reportDataCsv: SafeUrl;
+
+  constructor(private sanitizer: DomSanitizer) {}
+
+  @Input()
+  set reportData(data: any[][]) {
+    const joinedCsvArray =
+      data &&
+      'data:text/csv;charset=utf-8,' +
+        data
+          .map(
+            row =>
+              row &&
+              row
+                .map(v => {
+                  return v && typeof v === 'string' && v.indexOf(',') > -1 ? `"${v}"` : v;
+                })
+                .join(',')
+          )
+          .join('\r\n');
+    this.reportDataCsv = this.sanitizer.bypassSecurityTrustUrl(joinedCsvArray);
+  }
+
+  getFlightDates(flights: Flight[]): { startAt: Date; endAt: Date } {
+    return (
+      flights &&
+      flights.reduce(
+        (acc, flight) => {
+          let startAt = acc.startAt;
+          let endAt = acc.endAt;
+          if (flight.contractStartAt && (!acc.startAt || acc.startAt.valueOf() > flight.contractStartAt.valueOf())) {
```

Comment:
Yeah, they said they hadn't filled most of the contract dates out. It confused me that this is what they wanted in the report.

Ok, so the fields are independent... So what if there is a contract start but not a contract end date, but actual start and end date are filled out as those are required. What should it display then? Should the contract start date be mixed with the actual end date? Honestly it might make more sense to list both dates separately as Kali was suggesting yesterday.

Reply:
Good questions!

My perhaps bad assumption is: if you leave contract-start or contract-end blank, they "default" to the start or end dates.  At least as far as reporting is concerned!  So here i'd expect these to be sort of coalesced fields... `startAt: flight.contractStartAt || flight.startAt` (and the same with endAt).

But I would also be fine with just displaying them both.  More data, more clarity!